memgpt/server/server.py

```diff
@@ -848,67 +850,41 @@ def _agent_state_to_config(self, agent_state: AgentState) -> dict:
-
-        for agent_state, return_dict in zip(agents_states, agents_states_dicts):
-
-            # Get the agent object (loaded in memory)
-            memgpt_agent = self._get_or_load_agent(user_id=user_id, agent_id=agent_state.id)
-
-            # TODO remove this eventually when return type get pydanticfied
-            # this is to add persona_name and human_name so that the columns in UI can populate
-            # TODO hack for frontend, remove
-            # (top level .persona is persona_name, and nested memory.persona is the state)
-            # TODO: eventually modify this to be contained in the metadata
-            return_dict["persona"] = agent_state._metadata.get("persona", None)
-            return_dict["human"] = agent_state._metadata.get("human", None)
-
-            # Add information about tools
-            # TODO memgpt_agent should really have a field of List[ToolModel]
-            #      then we could just pull that field and return it here
-            return_dict["tools"] = [tool for tool in all_available_tools if tool.json_schema in memgpt_agent.functions]
-
-            # Add information about memory (raw core, size of recall, size of archival)
-            core_memory = memgpt_agent.memory
-            recall_memory = memgpt_agent.persistence_manager.recall_memory
-            archival_memory = memgpt_agent.persistence_manager.archival_memory
-            memory_obj = {
-                "core_memory": {section: module.value for (section, module) in core_memory.memory.items()},
-                "recall_memory": len(recall_memory) if recall_memory is not None else None,
-                "archival_memory": len(archival_memory) if archival_memory is not None else None,
-            }
-            return_dict["memory"] = memory_obj
+        agents_states = self.ms.list_agents(user_id=user_id, after=after, before=before, limit=limit, order=order)
+        agents_states = [agentstate_to_agentstatemodel(a_s) for a_s in agents_states]
 
-            # Add information about last run
-            # NOTE: 'last_run' is just the timestamp on the latest message in the buffer
-            # Retrieve the Message object via the recall storage or by directly access _messages
-            last_msg_obj = memgpt_agent._messages[-1]
-            return_dict["last_run"] = last_msg_obj.created_at
+        # For the UI listing panel, we need to add the following metadata:
+        # - tool breakdown (num base/core tools, num other)
+        # - last run data
+        # - memory stats (recall, archival)
+        # - number of sources
+        for agent_state in agents_states:
 
             # Add information about attached sources
             sources_ids = self.ms.list_attached_sources(agent_id=agent_state.id)
             sources = [self.ms.get_source(source_id=s_id) for s_id in sources_ids]
-            return_dict["sources"] = [vars(s) for s in sources]
+            agent_state.metadata["sources"] = [vars(s) for s in sources]
```

Comment:
this would also be better to place into the the `update_state` function - might need to add a `Agent.sources` variable

Reply:
TODO in next commit - strip these, client should instead to a 2nd/3rd GET to poll this information

Reply:
also moved out to separate rest routes

we should make an issue though for putting `sources` inside of agent?